tidy3d/components/lumped_element.py

```diff
@@ -107,24 +140,148 @@ def _voltage_axis_in_plane(cls, val, values):
-    as :class:`Medium2D` with the appropriate conductivity given their size and voltage axis."""
+    @cached_property
+    def to_structure(self) -> Structure:
+        """Converts the :class:`LumpedResistor` object to a :class:`Structure`
+        ready to be added to the :class:`Simulation`"""
+        conductivity = self._sheet_conductance
+        medium_dict = {
+            "tt": Medium(conductivity=conductivity),
+            "ss": Medium(conductivity=conductivity),
+        }
+        return Structure(
+            geometry=Box(size=self.size, center=self.center),
+            medium=Medium2D(**medium_dict),
+        )
 
-    resistance: pydantic.PositiveFloat = pydantic.Field(
+
+class CoaxialLumpedResistor(AbstractLumpedResistor):
+    """Class representing a coaxial lumped resistor. Lumped resistors are appended to the list of
+    structures in the simulation as :class:`Medium2D` with the appropriate conductivity given their
+    size and geometry."""
+
+    center: Coordinate = pydantic.Field(
+        (0.0, 0.0, 0.0),
+        title="Center",
+        description="Center of object in x, y, and z.",
+        units=MICROMETER,
+    )
+
+    outer_diameter: pydantic.PositiveFloat = pydantic.Field(
         ...,
-        title="Resistance",
-        description="Resistance value in ohms.",
-        unit=OHM,
+        title="Outer Diameter",
+        description="Diameter of the outer concentric circle.",
+        units=MICROMETER,
+    )
+
+    inner_diameter: pydantic.PositiveFloat = pydantic.Field(
+        ...,
+        title="Inner Diameter",
+        description="Diameter of the inner concentric circle.",
+        units=MICROMETER,
+    )
+
+    normal_axis: Axis = pydantic.Field(
+        ...,
+        title="Normal Axis",
+        description="Specifies the normal axis, which defines"
+        "the orientation of the circles making up the coaxial lumped element.",
     )
 
+    def to_mesh_overrides(self) -> List[MeshOverrideStructure]:
+        """Creates a suggested :class:`MeshOverrideStructure` list that could be added to the
+        :class:`Simulation`"""
+        # Create a mesh override when refinement is needed.
+        # An important use case is when a LumpedResistor is used with a LumpedPort
+        # The port is a flat surface, but when computing the port current,
+        # we'll eventually integrate the magnetic field just above and below
+        # this surface, so the mesh override needs to ensure that the mesh
+        # is fine enough not only in plane, but also in the normal direction.
+        # So in the normal direction, we'll make sure there are at least
+        # 2 cell layers above and below whose size is the same as the in-plane
+        # cell size in the override region. Also, to ensure that the port itself
+        # is aligned with a grid boundary in the normal direction, two separate
+        # override regions are defined, one above and one below the analytical
+        # port region.
```

Comment:
This refinement along the normal axis ensures that the H field is as close as possible to the sheet, thereby allowing for a more accurate approximation of the current? How does this affect accuracy?

Reply:
That is actually a great question :smile:! I tried a simulation of a microstrip with regular Rectangular lumped ports. Here are three sets of results where I have computed the root mean square error of the 4 S parameters:

```
# Refine Ports Off
# RMSE td:  0.11258103731991326
# RMSE td:  0.16233394675835855
# RMSE td:  0.1623539382006575
# RMSE td:  0.11795092811347076

# Refine Ports On with Normal Refinement (10 grid cells across port)
# RMSE td:  0.09013413380040426
# RMSE td:  0.11568887419087617
# RMSE td:  0.11568856206400055
# RMSE td:  0.09021164537686488

# Refine Ports On without Normal Refinement (10 grid cells across port)
# RMSE td:  0.10821387714324963
# RMSE td:  0.09544047779802187
# RMSE td:  0.09544046629736976
# RMSE td:  0.10825511559197067
```

So definitely port refinement can improve accuracy, but I think when refining along the normal axis it may be detrimental, or at least it does not improve the accuracy much. I think this is due to the self-inductance, which is related to the size and distribution of the current source and smaller size along the normal axis increases the self-inductance. It's affect on S parameters is also more pronounced at higher frequencies.

Shashwat originally implemented this functionality, which also helped in situations where the original grid did not exactly line up with the port center along the normal axis. Since then I improved how path integrals are performed and how ports are snapped to grid cell boundaries, so it might be less necessary.

But I think modifying this part of the code should be left for a different PR, what do you think?

Reply:
Thanks, @dmarek-flex! I raised up this question because the refinement along the normal direction introduces some computational costs, including an increased number of cells and a restricted timestep. It appears that this refinement doesn’t lead to improved accuracy, so could we consider reducing this refinement in the future? I have no concerns at this moment, all of these setup are solid. I just want to confirm that the final setup will not require intensive mesh refinement along the normal direction.